src/object_store/src/object/s3.rs

```diff
@@ -28,7 +29,7 @@ pub struct S3ObjectStore {
 
 #[async_trait::async_trait]
 impl ObjectStore for S3ObjectStore {
-    async fn upload(&self, path: &str, obj: Bytes) -> ObjectResult<()> {
+    async fn upload(&self, path: &str, obj: Vec<u8>) -> ObjectResult<()> {
```

Comment:
Does `ByteStream::from(Vec<u8>)` cause an extra copy?

Reply:
Initially, I thought there is not because `BytesStream::form(Vec<u8>)` just calls `Bytes::form(Vec<u8>)` and use `Vec<u8>` as `Bytes` underlying data buffer which is zero-cost. But after deeper investigation, I found it is not.

When converting `Vec<u8>` to `Bytes`, the `from` function called `Vec::into_boxed_slice` to unwrap the `Vec<u8>` and drop excess capacity (to make sure the excess capacity can be freed), which means if the size and capacity of the origin `Vec<u8>` are not the same, `shrink` will be called to reallocate buffer (nearly zero-cost, because the new layout must be smaller than the origin one, so there is no need to allocate a new buffer) and `ptr::copy_non_overlapping`.

```rust
unsafe fn shrink(
        &self,
        ptr: NonNull<u8>,
        old_layout: Layout,
        new_layout: Layout,
    ) -> Result<NonNull<[u8]>, AllocError> {
        debug_assert!(
            new_layout.size() <= old_layout.size(),
            "`new_layout.size()` must be smaller than or equal to `old_layout.size()`"
        );

        let new_ptr = self.allocate(new_layout)?;

        // SAFETY: because `new_layout.size()` must be lower than or equal to
        // `old_layout.size()`, both the old and new memory allocation are valid for reads and
        // writes for `new_layout.size()` bytes. Also, because the old allocation wasn't yet
        // deallocated, it cannot overlap `new_ptr`. Thus, the call to `copy_nonoverlapping` is
        // safe. The safety contract for `dealloc` must be upheld by the caller.
        unsafe {
            ptr::copy_nonoverlapping(ptr.as_ptr(), new_ptr.as_mut_ptr(), new_layout.size());
            self.deallocate(ptr, old_layout);
        }

        Ok(new_ptr)
    }
```

I did a benchmark to see if Rust optimized it when there is no actual allocation, but sadly, I found it didn't.

https://gist.github.com/MrCroxx/f2594270180b8bde6f74b7e110790b44

Copying around 1GiB data requires around 1ms. But I think we can tolerant the overhead in our case.